pingsource-mt-adapter dispatcher panics during reconciler-tests

While combing through the test artifacts output by https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative_eventing/5419/pull-knative-eventing-reconciler-tests/1395112946597629952 I saw that there is a panic coming from the pingsource-mt-adapter dispatcher:

```
2021-05-19T20:52:13.432Z  debug runtime/panic.go:965  Finished sending cloudevent id: 2248dfa9-6634-4db9-a081-bac62ec14347  {"commit": "38199f7"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1750e2e]

goroutine 232 [running]:
knative.dev/eventing/pkg/adapter/v2.(*client).reportMetrics(0xc0001731a0, 0x1e71198, 0xc0007d1710, 0x1e95180, 0xc0000d2e00, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	knative.dev/eventing/pkg/adapter/v2/cloudevents.go:185 +0x48e
knative.dev/eventing/pkg/adapter/v2.(*client).Send(0xc0001731a0, 0x1e71198, 0xc0007d1710, 0x1e95180, 0xc0000d2e00, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	knative.dev/eventing/pkg/adapter/v2/cloudevents.go:122 +0x125
knative.dev/eventing/pkg/adapter/mtping.(*cronJobsRunner).cronTick.func1()
	knative.dev/eventing/pkg/adapter/mtping/runner.go:136 +0x579
github.com/robfig/cron/v3.FuncJob.Run(0xc0003a9500)
	github.com/robfig/cron/v3@v3.0.1/cron.go:136 +0x25
github.com/robfig/cron/v3.(*Cron).startJob.func1(0xc000157860, 0x1e3d740, 0xc0003a9500)
	github.com/robfig/cron/v3@v3.0.1/cron.go:312 +0x65
created by github.com/robfig/cron/v3.(*Cron).startJob
	github.com/robfig/cron/v3@v3.0.1/cron.go:310 +0x73
```

--- a/pkg/adapter/v2/cloudevents.go
+++ b/pkg/adapter/v2/cloudevents.go
@@ -181,8 +181,9 @@ func (c *client) reportMetrics(ctx context.Context, event cloudevents.Event, res
 			var res *http.Result
 			if !cloudevents.ResultAs(retryResult, &res) {
 				c.reportError(reportArgs, result)
+			} else {
+				c.reporter.ReportRetryEventCount(reportArgs, res.StatusCode)
 			}
-			c.reporter.ReportRetryEventCount(reportArgs, res.StatusCode)
 		}
 	}
 }

--- a/pkg/adapter/v2/cloudevents_test.go
+++ b/pkg/adapter/v2/cloudevents_test.go
@@ -64,11 +64,13 @@ func TestNewCloudEventsClient_send(t *testing.T) {
 		timeout        int
 		wantErr        bool
 		wantRetryCount bool
+		wantEventCount int
 	}{
 		"timeout": {timeout: 13},
 		"none":    {},
 		"send": {
-			event: demoEvent(),
+			event:          demoEvent(),
+			wantEventCount: 1,
 		},
 		"send with retries": {
 			event: func() *cloudevents.Event {
@@ -79,12 +81,14 @@ func TestNewCloudEventsClient_send(t *testing.T) {
 				return &event
 			}(),
 			wantRetryCount: true,
+			wantEventCount: 1,
 		},
 		"send with ceOverrides": {
 			event: demoEvent(),
 			ceOverrides: &duckv1.CloudEventOverrides{Extensions: map[string]string{
 				"foo": "bar",
 			}},
+			wantEventCount: 1,
 		},
 		"send fails": {
 			event: func() *cloudevents.Event {
@@ -94,7 +98,8 @@ func TestNewCloudEventsClient_send(t *testing.T) {
 				event.SetType("unit.sendFail")
 				return &event
 			}(),
-			wantErr: true,
+			wantErr:        true,
+			wantEventCount: 1,
 		},
 		"not a http result": {
 			event: func() *cloudevents.Event {
@@ -104,7 +109,19 @@ func TestNewCloudEventsClient_send(t *testing.T) {
 				event.SetType("unit.wantErr")
 				return &event
 			}(),
-			wantErr: true,
+			wantErr:        true,
+			wantEventCount: 1,
+		},
+		"retry that is not a http result": {
+			event: func() *cloudevents.Event {
+				event := cloudevents.NewEvent()
+				event.SetID("abc-123")
+				event.SetSource("unit/test")
+				event.SetType("unit.nonHttpRetry")
+				return &event
+			}(),
+			wantErr:        false,
+			wantEventCount: 2,
 		},
 	}
 
@@ -181,7 +198,7 @@ func TestNewCloudEventsClient_send(t *testing.T) {
 					t.Fatal(err)
 				}
 				validateSent(t, innerClient, tc.event.Type())
-				validateMetric(t, got.reporter, 1, tc.wantRetryCount)
+				validateMetric(t, got.reporter, tc.wantEventCount, tc.wantRetryCount)
 			} else {
 				validateNotSent(t, innerClient)
 			}

--- a/pkg/adapter/v2/test/test_client.go
+++ b/pkg/adapter/v2/test/test_client.go
@@ -86,6 +86,10 @@ func (c *TestCloudEventsClient) Send(ctx context.Context, out event.Event) proto
 		return errors.New("totally not an http result")
 	} else if eventData.Type == "unit.sendFail" {
 		return http.NewResult(400, "%w", protocol.ResultNACK)
+	} else if eventData.Type == "unit.nonHttpRetry" {
+		var attempts []protocol.Result
+		attempts = append(attempts, errors.New("totally not an http result"))
+		return http.NewRetriesResult(http.NewResult(200, "%w", protocol.ResultACK), 1, time.Now(), attempts)
 	}
 	return http.NewResult(200, "%w", protocol.ResultACK)
 }

